### PR TITLE
Do not allow uncompressed pubkey in `wpkh(KEY)` output descriptor

### PR DESCRIPTION
**Pull request by kiminuo**
> wpkh(KEY) (not inside wsh): P2WPKH output for the given compressed pubkey.

### CODE REVIEW

**Comment by kiminuo** [comment]
Fuzz crash is thanks to #89. Hopefully.

File: src/miniscript/astelem.rs [hunk 1]
```diff
@@ -364,9 +364,9 @@ where
             }
         }
         let mut unwrapped = match (frag_name, top.args.len()) {
-            ("pk_k", 1) => {
-                expression::terminal(&top.args[0], |x| Pk::from_str(x).map(Terminal::PkK))
-            }
+            ("pk_k", 1) => expression::terminal(&top.args[0], |x| {
+                str::FromStr::from_str(x).map(Terminal::PkK)
```
[the remark this answers is not in the document]
Comment (kiminuo) [reply src/miniscript/astelem.rs:368]: Modified

File: src/lib.rs [hunk 2]
```diff
@@ -126,6 +132,17 @@ impl MiniscriptKey for bitcoin::PublicKey {
         self.write_into(&mut engine);
         hash160::Hash::from_engine(engine)
     }
+
+    fn from_str(s: &str, compressed_only: bool) -> Result<Self, Self::Err> {
+        let r: bitcoin::PublicKey = str::FromStr::from_str(s)?;
+        if compressed_only && !r.compressed {
+            Err(Base58(base58::Error::Other(
+                "Uncompressed pubkeys are not allowed".to_owned(),
```
[the remark this answers is not in the document]
Comment (kiminuo) [reply src/lib.rs]: Yes. Modified.

File: src/policy/concrete.rs [hunk 3]
```diff
@@ -381,7 +381,9 @@ where
             }
         }
         match (frag_name, top.args.len() as u32) {
-            ("pk", 1) => expression::terminal(&top.args[0], |pk| Pk::from_str(pk).map(Policy::Key)),
+            ("pk", 1) => expression::terminal(&top.args[0], |pk| {
```
[the remark this answers is not in the document]
Comment (kiminuo) [reply src/policy/concrete.rs:384]: I'm not sure I follow. Could you elaborate on that?

File: src/miniscript/astelem.rs (hunk 1 above)
[the remark this answers is not in the document]
Comment (kiminuo) [reply src/miniscript/astelem.rs]: Changed

**Comment by kiminuo** [comment]
> You would have to pass the compressed flag from the Terminal::from_tree().

I'll just rephrase what you said to make sure I understand correctly. So we have fixed `wpkh` - line: 

https://github.com/kiminuo/rust-miniscript/blob/f337f249ee545257676634e56e17173e67218666/src/descriptor/mod.rs#L408

and now we would like to continue with `wsh` here:

https://github.com/kiminuo/rust-miniscript/blob/f337f249ee545257676634e56e17173e67218666/src/descriptor/mod.rs#L404

where we would like to pass `compressed_only=true`. Meaning, one would need to modify `Miniscript::from_tree(arg)` to `Miniscript::from_tree(arg, compressed_only)`. Right? Is that what you suggest?

**Comment by kiminuo** [comment]
In the mean time, I have added a new commit that should address some of your suggestions. The goal is to squash the commits when everything is done.

**Comment by sanket1729** [comment]
On second thought, I think maybe altering the trait for FromTree might not be a good idea as this will result in passing the flag in all recursive calls of `from_tree()` . 

Maybe there is a better way than doing this, but I can't think as of now. I am wondering if selectively supporting 65 bytes keys for non-segwit is a good idea given the code changes. @apoelstra, what are your opinions on this?

**Comment by apoelstra** [comment]
I wouldn't be opposed to alterning the trait. It is annoying and ugly but not *overly* so.

What I'd really like though is to somehow use the type system to eliminate these checks entirely. Like, to make it impossible to construct a segwit Miniscript with 65-byte keys. But this might not be (cleanly) possible in Rust.

**Comment by apoelstra** [comment]
Consider also that in segwit v1 we will lose the `ThreshM` node (and gain some new one). I wonder if we should think about a more general way to split out Miniscripts for the different descriptor contexts.

**Comment by sanket1729** [comment]
Hey @kiminuo , sorry about this. We decided to overrule this approach with a generic script context parameter that should allow us for an easier transition to tapscript.

**Comment by kiminuo** [comment]
No problem :)

**Comment by sanket1729** [comment]
Closing in favor of #97

**Pull request closed without merging**